custom/ilsgateway/api.py

```diff
@@ -316,14 +319,83 @@ def product_sync(self, ilsgateway_product):
                     product.save()
         return product
 
-    def web_user_sync(self, ilsgateway_webuser):
-        web_user = super(ILSGatewayAPI, self).web_user_sync(ilsgateway_webuser)
-        if not web_user:
-            return None
-        dm = web_user.get_domain_membership(self.domain)
+    def _create_web_user(self, email, ilsgateway_webuser, location_id, user_dict):
+        user = WebUser.create(domain=None, username=email,
+                              password=ilsgateway_webuser.password, email=ilsgateway_webuser.email.lower(),
+                              **user_dict)
+        user.add_domain_membership(self.domain, location_id=location_id)
+        dm = user.get_domain_membership(self.domain)
         dm.role_id = UserRole.get_read_only_role_by_domain(self.domain).get_id
-        web_user.save()
-        return web_user
+        user.save()
+        return user
+
+    def _edit_web_user(self, user, location_id):
+        read_only_role = UserRole.get_read_only_role_by_domain(self.domain).get_id
+        if self.domain not in user.get_domains():
+            user.add_domain_membership(self.domain, location_id=location_id, role_id=read_only_role)
+            user.save()
+        else:
+            dm = user.get_domain_membership(self.domain)
+            if dm.role_id != read_only_role or dm.location_id != location_id:
+                dm.role_id = read_only_role
+                dm.location_id = location_id
+                user.save()
+
+    def web_user_sync(self, ilsgateway_webuser):
+        email = ilsgateway_webuser.email.lower()
+        if not email:
+            try:
+                validate_email(ilsgateway_webuser.username)
+                email = ilsgateway_webuser.username
+            except ValidationError:
+                return None
+        user = WebUser.get_by_username(email)
+        user_dict = {
+            'first_name': ilsgateway_webuser.first_name,
+            'last_name': ilsgateway_webuser.last_name,
+            'is_active': ilsgateway_webuser.is_active,
+            'password_hashed': True,
+        }
+        location_id = None
+        if ilsgateway_webuser.location:
+            try:
+                sql_location = SQLLocation.objects.get(domain=self.domain, external_id=ilsgateway_webuser.location)
+                location_id = sql_location.location_id
+            except SQLLocation.DoesNotExist:
+                location_id = None
+
+        if user is None:
+            try:
+                ils_sql_web_user = ILSGatewayWebUser.objects.get(username=ilsgateway_webuser.username)
```

Comment:
Unfortunately they can change all fields. It seems that I need to change API to return user id and this change should be based on user id.